IMR restore feature implementation

keyonjie opened the pull request:
This simply series implement the IMR restoring feature from the Linux side.

It is verified on cAVS platforms.

After switching to IMR restoring, the resuming boot time is decreased significantly:
e.g.

keyonjie commented:
The corresponding FW PR here: https://github.com/thesofproject/sof/pull/4320, though it is backward compatible so the 2 PRs can be merged separately.

@plbossart @ranj063 @kv2019i @bardliao please help to review.

The change here might be invasive as I don't add a kconfig item to enable/disable this new IMR restore feature, we can discuss if making it configurable is needed.

plbossart commented on include/sound/sof/info.h:
```diff
@@ -25,6 +25,7 @@
 #define SOF_IPC_INFO_LOCKS		BIT(1)
 #define SOF_IPC_INFO_LOCKSV		BIT(2)
 #define SOF_IPC_INFO_GDB		BIT(3)
+#define SOF_IPC_INFO_IMR_RESTORE	BIT(4)
```
ABI change?

keyonjie replied:
> ABI change?

It is backward compatible, e.g. all FW/driver combination works, so I think we don't need ABI bump here:
old FW + old driver, no IMR restore.
old FW + new driver, no IMR restore.
new FW + old driver, driver FW purge IPC and re-downloading, no IMR restore.
new FW + new driver, IMR restore.

kv2019i replied:
@keyonjie Yes, but it's still extending the public interface so
"Increment MINOR version if you add backwards compatible features or
 changes. PATCH should be reset to 0." applies. https://github.com/thesofproject/sof/blob/main/src/include/kernel/abi.h

keyonjie replied:
> ment MINOR version if you add backwards compatible features or
> changes. PATCH should be reset to 0.

OK, let me bump it to 3.19.0 then.

plbossart commented on sound/soc/sof/intel/hda-loader.c:
```diff
@@ -335,6 +335,56 @@ int hda_dsp_cl_boot_firmware_iccmax(struct snd_sof_dev *sdev)
+	 * enabled before accessing the SoundWire SHIM registers
+	 */
+	if (!sdev->first_boot)
+		hda_sdw_process_wakeen(sdev);
```
lines 354..383 look like copy/paste. can we avoid this please? At the very least introduce helpers for SSP and SoundWire.

keyonjie replied:
> lines 354..383 look like copy/paste. can we avoid this please? At the very least introduce helpers for SSP and SoundWire.

Um, I thought about this, then I didn't want to change the sequence (e.g. to move SSP/SDW to the last) to introduce unexpected risk and that's why you see this...

Since you are confirming the SSP/SDW can be moved to the very last, let me share that part to avoid copy/paste here.

plbossart replied:
I am not confirming anything wrt SSP/SDW, you need to keep the same order as before, but call a helper. Having duplicated code and comments isn't good.

keyonjie replied:
> I am not confirming anything wrt SSP/SDW, you need to keep the same order as before, but call a helper. Having duplicated code and comments isn't good.

got you.

plbossart commented on sound/soc/sof/intel/hda-loader.c:
```diff
@@ -335,6 +335,56 @@ int hda_dsp_cl_boot_firmware_iccmax(struct snd_sof_dev *sdev)
+	 * is initialized successfully, which ensures power rails are
+	 * enabled before accessing the SoundWire SHIM registers
+	 */
+	if (!sdev->first_boot)
```
what is the definition of 'first_boot' if we don't re-initialize the firmware?

keyonjie replied:
> what is the definition of 'first_boot' if we don't re-initialize the firmware?

We are only not re-downloading the firmware, the first/second/... FW boot concept are kept intact?

plbossart replied:
my point was: be careful and check for all usages of 'first boot'. In the past we would download and authenticate the firmware at every resume, now only at first boot. It's a legitimate question to revisit all usages of first_boot to make sure we don't forget about something.

keyonjie replied:
> my point was: be careful and check for all usages of 'first boot'. In the past we would download and authenticate the firmware at every resume, now only at first boot. It's a legitimate question to revisit all usages of first_boot to make sure we don't forget about something.

Got you, let me list all of those usages below to relax us about that.

The first_boot is initialize to be 'true' and changed to 'false' once the probing succeed in core.c, all remain usages of 'first boot' is reading:

in hda-loader.c, 3 usages:
hda-loader.c:406, hda_sdw_process_waken() called for resuming boot, kept unchanged in my PR.
hda-loader.c:463, 487, hda_sdw_startup() called at _post_fw_run_() for first boot only, I think we don't need it for resuming case, even we don't re-download the FW, @bardliao @RanderWang @plbossart can you help to confirm this?

in loader.c, we basically use first_boot to decide if requesting firmware, creating debugfs and parsing fw_ready message is needed, we still don't need those (as previous) with the change here.

in pm.c:100, first_boot is used to check if sof_resume() should be performed, if the first_boot is true, it means the probing was not succeed so sof_resume() still no needed.

in trace.c:485, first_boot is used for creating trace related debugfs entries, we don't need create them for resuming boot, this is not changed.

```
$ grep first_boot sound/soc/sof/* -rn
sound/soc/sof/core.c:222:       sdev->first_boot = false;
sound/soc/sof/core.c:270:       sdev->first_boot = true;
sound/soc/sof/core.c:303:       sdev->first_boot = true;
sound/soc/sof/intel/hda-loader.c:406:   if (!sdev->first_boot)
sound/soc/sof/intel/hda-loader.c:463:   if (sdev->first_boot) {
sound/soc/sof/intel/hda-loader.c:487:   if (sdev->first_boot) {
sound/soc/sof/loader.c:66:      if (sdev->first_boot) {
sound/soc/sof/loader.c:190:     if (sdev->first_boot)
sound/soc/sof/loader.c:230:                     if (sdev->first_boot && elem->value)
sound/soc/sof/loader.c:495:     if (!sdev->first_boot)
sound/soc/sof/loader.c:795:     if (sdev->first_boot) {
sound/soc/sof/pm.c:100: if (sdev->first_boot)
sound/soc/sof/sof-priv.h:396:   bool first_boot;
sound/soc/sof/trace.c:485:      if (sdev->first_boot) {

```

plbossart commented on sound/soc/sof/intel/hda-loader.c:
```diff
@@ -345,6 +395,12 @@ int hda_dsp_cl_boot_firmware(struct snd_sof_dev *sdev)
+	if ((sdev->fw_ready.flags & SOF_IPC_INFO_IMR_RESTORE) &&
+	    !sdev->first_boot) {
+		dev_info(sdev->dev, "info: IMR restore supported, booting from IMR directly\n");
+		return hda_dsp_boot_imr(sdev);
```
what happens if we override this IMR_RESTORE configuration. Can we still go ahead and proceed with the default path?

This should be an opt-in solution IMHO. Changing the flows will mechanically lead to bugs, so by default users who upgrade the firmware should not see a change on how the DSP is managed.

keyonjie replied:
> what happens if we override this IMR_RESTORE configuration. Can we still go ahead and proceed with the default path?

Yes, if the driver choose to re-downloading, even the IMR_RESTORE is declared support from the fw_ready, we will still go ahead with today's path.

> 
> This should be an opt-in solution IMHO. Changing the flows will mechanically lead to bugs, so by default users who upgrade the firmware should not see a change on how the DSP is managed.

Um, that's what I commented above, do you suggest a kconfig item or a module param here?

kv2019i replied:
@plbossart This can also fix issues, so not sure if this should be opt-in. If tests look good, we could just make this the default and offer an opt-out in case an issue is hit.

plbossart replied:
@kv2019i yes, that's what I meant, opt-out/disable similar to pm_runtime or dynamic pipelines.

keyonjie replied:
So let me add a kconfig item and the check will be like this:
```c
if ((sdev->fw_ready.flags & SOF_IPC_INFO_IMR_RESTORE) &&
            !CONFIG_SND_SOC_SOF_DEBUG_DISABLE_IMR_RESTORE &&
	    !sdev->first_boot) {
```

kv2019i replied:
@keyonjie @plbossart I'd vote for a module option. If this causes a regression, user can easily add a modprobe.d option, but having to roll out a custom kernel for a specific system is too difficult. And we have to too many Kconfig options. If there are many, we should have a quirk table in driver to make the default sane for most systems (as the module option is not scalable either).

plbossart replied:
agree with @kv2019i we already have sof_debug quirks as a module parameter, let's add one bit to disable IMR reload.

plbossart commented on sound/soc/sof/intel/hda-loader.c:
```diff
@@ -335,6 +335,56 @@ int hda_dsp_cl_boot_firmware_iccmax(struct snd_sof_dev *sdev)
+	 * is initialized successfully, which ensures power rails are
+	 * enabled before accessing the SoundWire SHIM registers
+	 */
+	if (!sdev->first_boot)
```
On a similar note, when is the firmware preserved in IMR?
D3/S0 aka pm_runtime_suspend or RTD3?
D3/S3 aka system suspend/hibernate?

keyonjie replied:
> On a similar note, when is the firmware preserved in IMR?
> D3/S0 aka pm_runtime_suspend or RTD3?
> D3/S3 aka system suspend/hibernate?

Per my understanding, it is preserved in case the DDR is with power, so yes for both D3/S0 and D3/S3. 

I am not quite sure about the S4 (Suspend-to-Disk) case though, what do you think about that?

plbossart replied:
That's for you to find out @keyonjie, and it goes back to my question on 'first-boot'. It's rather important to know when the IMR is preserved and when it's not.

keyonjie replied:
> That's for you to find out @keyonjie, and it goes back to my question on 'first-boot'. It's rather important to know when the IMR is preserved and when it's not.

Since I can't find a platform that support S4 hibernation at my end, will try this later.

plbossart reviewed (changes requested):
Good idea to finally follow the recommended flows, but comments below.
Thanks @keyonjie for starting this.

kv2019i commented on include/sound/sof/info.h:
```diff
@@ -25,6 +25,7 @@
 #define SOF_IPC_INFO_LOCKSV		BIT(2)
 #define SOF_IPC_INFO_GDB		BIT(3)
+#define SOF_IPC_INFO_IMR_RESTORE	BIT(4)
 
```
@dbaluta Do you see any need for a fast-boot flag indiciated by FW? 

This could be perhaps SOF_IPC_INFO_FASTBOOT

... if defined, FW supports retaining some state even if if suspended. The actual details would be vendor specific, but we could use a common flag in FW_READY that all vendors could use.

ranj063 replied:
and the second line is a bit confusing right? you're not showing how to choose the feature or not just yet. So its better to just say: "If the bit is set, the driver can skip downloading the firmware again during system resume or runtime resume"

kv2019i reviewed (changes requested):
Excellent @keyonjie ! I do think we need to to bump the minor version and we'd need to come up some way to rename the flag name. Either make it generic (IMR is Intel specific) or use some vendor specific method to pass info (ext-manifest). I think this is generally useful indication, so maybe we can just rename the flag.

lyakh commented on sound/soc/sof/intel/hda-loader.c:
```diff
@@ -335,6 +335,56 @@ int hda_dsp_cl_boot_firmware_iccmax(struct snd_sof_dev *sdev)
+	init_waitqueue_head(&sdev->boot_wait);
+
+	/* power up & unstall/run core 0 to run the firmware */
+	ret = hda_dsp_enable_core(sdev, BIT(0));
```
`BIT(SOF_DSP_PRIMARY_CORE)`

lyakh commented on sound/soc/sof/intel/hda-loader.c:
```diff
@@ -335,6 +335,56 @@ int hda_dsp_cl_boot_firmware_iccmax(struct snd_sof_dev *sdev)
+	/* init for booting wait */
+	init_waitqueue_head(&sdev->boot_wait);
+
+	/* power up & unstall/run core 0 to run the firmware */
```
s/core 0/the primary core/

lyakh commented on sound/soc/sof/intel/hda-loader.c:
```diff
@@ -335,6 +335,56 @@ int hda_dsp_cl_boot_firmware_iccmax(struct snd_sof_dev *sdev)
+	hda_dsp_ipc_int_enable(sdev);
+
+	/*
+	 * When a SoundWire link is in clock stop state, a Slave
```
IIRC the standard vocabulary has been updated?

keyonjie replied:
> IIRC the standard vocabulary has been updated?

Not quite sure if it has been updated for SDW part, @plbossart should I s/Slave/secondary ?

plbossart replied:
we'll to update later.

keyonjie commented:
@kv2019i @plbossart @lyakh updated with most comments addressed.
Kconfig item not added as I want to get the alignment on that first.
S4 supporting not verified yet.

kv2019i commented on sound/soc/sof/intel/hda-loader.c:
```diff
@@ -345,6 +398,12 @@ int hda_dsp_cl_boot_firmware(struct snd_sof_dev *sdev)
 
+	if ((sdev->fw_ready.flags & SOF_IPC_INFO_D3_PERSISTENT) &&
+	    !sdev->first_boot) {
+		dev_info(sdev->dev, "info: IMR restore supported, booting from IMR directly\n");
```
This will a very frequent and normal flow, so dev_info() is too verbose. Maybe dev_dbg() for this.

kv2019i reviewed (commented):
Looks good now. Only thing left is for us to agree on the opt-out mechanism. Currently patch has none, @keyonjie has proposed Kconfig and now made inline comment that I'd  prefer a module parameter.

kv2019i commented:
Removed "[RFC]" as this is no longer a draft.

keyonjie commented:
@kv2019i @plbossart module option added.

keyonjie commented:
SOFCI TEST

kv2019i commented on sound/soc/sof/sof-priv.h:
```diff
@@ -30,6 +30,9 @@
 							*/
+#define SOF_DBG_DISABLE_D3_PERSISTENT	BIT(5)	/* disable D3 persistent feature
+						 * and FW reload needed
+						 */
```
This is maybe details, but as we already have SBF_DBG_DYNAMIC_PIPELINES_ENABLE, it feels inconsistent to have DISABLE for this feature. I wonder if "IGNORE_D3_PERSISTENT" would be more appropriate as this feature depends on FW support in the end, but if driver ignores the bit, we always do a cold boot.

SOF_DBG_FORCE_COLD_BOOT is another option, but probably better use D3_PERSISTENT as that's the FW_READY flag, so it's obvious that the two are connected.

Just suggestions, I can live with this name. I do think the comment needs to be reworded a bit. Something like: 
```
/* ignore DSP D3 persistent capability\n and always cold boot */
```

keyonjie replied:
Um, it's not like the system 'cold boot' concept here, the IMR is still powered on during the DSP-reboot here even though the IMR restore feature is not used, but yes, "IGNORE_D3_PERSISTENT" sounds better to me.

kv2019i replied:
Ack @keyonjie , maybe drop the "cold boot" term, it's not really accurate. It's just a different boot flow.

kv2019i reviewed (changes requested):
Only minor nitpick on the module param name and description. Otherwise looks good.

plbossart commented on include/sound/sof/info.h:
```diff
@@ -25,6 +25,7 @@
 #define SOF_IPC_INFO_LOCKS		BIT(1)
 #define SOF_IPC_INFO_LOCKSV		BIT(2)
 #define SOF_IPC_INFO_GDB		BIT(3)
+#define SOF_IPC_INFO_D3_PERSISTENT	BIT(4)
```
@keyonjie your commit message is way too specific. 

D3_PERSISTENT could also be used for Baytrail where the  DSP is not power gated. This bit should really mean that downloading firmware after a D3 transition is not required, whether the DSP remained powered or the memory was power-gated and restored is a platform detail that the kernel does not need to know, does it?

plbossart replied:
Also this has to be 3.20, I took 3.19 for SSP changes.

keyonjie replied:
I am not good at this naming stuff, since it has gone with this on the FW side as a FW_READY flag, let's just align to it?

For the ABI version, can we use 3.19 for both updates? It is a little bit screwed as the FW side has taken this D3_PERSISTENT as 3.19 and driver side took SSP changes as 3.19.
To fix this, as before can we use one minor version bump for several updates in a bunch? @lgirdwood FYI.

kv2019i replied:
@plbossart @keyonjie we can use 3.19 for both 3.19 already went out with a release (and I thin that's not the case, the tagged v1.8 has ABI3.18.1), so 3.19 is for SOF v1.9.

plbossart replied:
@kv2019i I don't think 3.19 is out, I bumped the number in my SSP changes and it's not merged yet.

plbossart commented on sound/soc/sof/intel/hda-loader.c:
```diff
@@ -26,6 +26,44 @@
+	if (!sdev->first_boot)
+		hda_sdw_process_wakeen(sdev);
+}
+
```
first add the helpers in the existing code, then add the IMR stuff in a second patch.

keyonjie replied:
Sure, let me change.

plbossart commented on sound/soc/sof/core.c:
```diff
@@ -22,6 +22,7 @@
 int sof_core_debug;
 module_param_named(sof_debug, sof_core_debug, int, 0444);
 MODULE_PARM_DESC(sof_debug, "SOF core debug options (0x0 all off)");
+EXPORT_SYMBOL(sof_core_debug);
```
not sure if this EXPORT is something we want to do? @kv2019i @ranj063 what do you think?

keyonjie replied:
> not sure if this EXPORT is something we want to do? @kv2019i @ranj063 what do you think?

This is needed in case we want to use the sof_core_debug from another module.

kv2019i replied:
@plbossart wondered about the same. Easy fix would be use a different parameter and put it under "intel/", but now that we use a generic debug flag, also the option should be here. The variable is in sof-priv.h and there are some examples of plain global variables exported (like sound_debugfs_root), so I guess this is ok.

plbossart replied:
maybe mirror the parameter into a structure (sdev) and test the structure member?

ranj063 commented:
@keyonjie will this enable IMR restore for both rintime PM and system PM? I think the recommendation is to use this feature only for runtime PM right?

keyonjie commented:
> @keyonjie will this enable IMR restore for both rintime PM and system PM? I think the recommendation is to use this feature only for runtime PM right?

yes, for both. otherwise the issue like https://github.com/thesofproject/linux/issues/2805 will still be reproduced with suspend/resume stress test.

ranj063 commented:
> > @keyonjie will this enable IMR restore for both rintime PM and system PM? I think the recommendation is to use this feature only for runtime PM right?
> 
> yes, for both. otherwise the issue like #2805 will still be reproduced with suspend/resume stress test.

what re the implications on power for system suspend with this?

keyonjie commented:
> > > @keyonjie will this enable IMR restore for both rintime PM and system PM? I think the recommendation is to use this feature only for runtime PM right?
> > 
> > 
> > yes, for both. otherwise the issue like #2805 will still be reproduced with suspend/resume stress test.
> 
> what re the implications on power for system suspend with this?

I think nothing is changed on power consumption for system suspend with this, but yes we need measured data to prove this.

ranj063 commented:
> I think nothing is changed on power consumption for system suspend with this, but yes we need measured data to prove this.

Is this supposed to work as intended even during suspend to S3?

keyonjie commented:
> > I think nothing is changed on power consumption for system suspend with this, but yes we need measured data to prove this.
> 
> Is this supposed to work as intended even during suspend to S3?

Think so, I verified this on TGL.
For S4, I am not quite sure as I don't have platform that I can try S4, anybody who can try it will be really appreciated.

BTW, let's do more testing before merging this PR. I will ask for a full round on all available platforms, especially with those suspend-resume related cases.

keyonjie commented:
SOFCI TEST

keyonjie commented:
@plbossart @ranj063 @kv2019i the CI captures several DSP reboot failures, so [DO NOT MERGE] it for the moment please.

plbossart commented:
> > > > @keyonjie will this enable IMR restore for both rintime PM and system PM? I think the recommendation is to use this feature only for runtime PM right?
> > > 
> > > 
> > > yes, for both. otherwise the issue like #2805 will still be reproduced with suspend/resume stress test.
> > 
> > 
> > what re the implications on power for system suspend with this?
> 
> I think nothing is changed on power consumption for system suspend with this, but yes we need measured data to prove this.

I can see how suspend-to-RAM would work, as long as the memory remains in self-refresh mode the context would be preserved.
Suspend-to-disk is interesting, would IMR be saved to the disk?

keyonjie commented:
SOFCI TEST

keyonjie commented:
@ranj063 it's interesting that this fails on HDA dynamic pipeline platforms, and after check-ipc-flood case, the FW can't boot from IMR anymore.

plbossart commented:
> @ranj063 it's interesting that this fails on HDA dynamic pipeline platforms, and after check-ipc-flood case, the FW can't boot from IMR anymore.

why was this merged in the firmware then if the functionality is still problematic? Confused.

keyonjie commented:
> > @ranj063 it's interesting that this fails on HDA dynamic pipeline platforms, and after check-ipc-flood case, the FW can't boot from IMR anymore.
> 
> why was this merged in the firmware then if the functionality is still problematic? Confused.

The issue was not found without this CI widely validation.

plbossart commented:
> > > @ranj063 it's interesting that this fails on HDA dynamic pipeline platforms, and after check-ipc-flood case, the FW can't boot from IMR anymore.
> > 
> > 
> > why was this merged in the firmware then if the functionality is still problematic? Confused.
> 
> The issue was not found without this CI widely validation.

Of course. CI tested the firmware in isolation without the kernel patch, so there's no change in terms of the flow. The kernel ignores the new IMR stuff.
So when the kernel does change the flow every breaks.

That tells me we did not try the combined firmware + kernel patches together.

lgirdwood commented:
> > > > @ranj063 it's interesting that this fails on HDA dynamic pipeline platforms, and after check-ipc-flood case, the FW can't boot from IMR anymore.
> > > 
> > > 
> > > why was this merged in the firmware then if the functionality is still problematic? Confused.
> > 
> > 
> > The issue was not found without this CI widely validation.
> 
> Of course. CI tested the firmware in isolation without the kernel patch, so there's no change in terms of the flow. The kernel ignores the new IMR stuff.
> So when the kernel does change the flow every breaks.
> 
> That tells me we did not try the combined firmware + kernel patches together.

FW was merged to to get more more testing coverage without the kernel change on all platforms (which is a valid use case with all related other FWs).

@keyonjie FW can be reverted or incrementally updated if you have a fix today.

keyonjie commented:
@lgirdwood @plbossart I am debugging on this, the draft result now is that the lack of DSP CORE ref_cnt in the driver side make this fail on dynamic pipeline platforms, please hold on, I will come with more information today.

plbossart commented:
> @lgirdwood @plbossart I am debugging on this, the draft result now is that the lack of DSP CORE ref_cnt in the driver side make this fail on dynamic pipeline platforms, please hold on, I will come with more information today.

This is getting out of control. We can't possibly deal with multi-core, dynamic pipelines and IMR at the same time. I realize there are side-effects and interaction but if we can't complete one round of validation for an individual capability across platforms it's not productive to add more features on top. We have to define baby steps instead of having both kernel and firmware changes across the board.

keyonjie commented:
@plbossart @ranj063 @kv2019i the dynamic pipeline fails with IPC flooding case running first issue addressed and updated, for APL failure, as mentioned on https://github.com/thesofproject/sof/pull/4386, the PR suggest to disable the feature on cAVS 1.5 and revisit it once it is required later.

keyonjie commented:
@plbossart @kv2019i I added more platforms (includes the strict TGLU_RVP_NOCODEC) coverage and triggered a validation with combine this PR and the SOF thesofproject/sof#4386, the result looks quite good. Please refer to the CI test plan result 4814.

plbossart commented:
> @plbossart @kv2019i I added more platforms (includes the strict TGLU_RVP_NOCODEC) coverage and triggered a validation with combine this PR and the SOF [thesofproject/sof#4386](https://github.com/thesofproject/sof/pull/4386), the result looks quite good. Please refer to the CI test plan result 4814.

@keyonjie it would be good if your tests used the *same* devices as the regular daily tests. You seem to have excellent results indeed, but added and removed platforms, so we can't do an apples-to-apples comparison.
Missing platforms include: 

TGLH_0A5E_SDW | EHL_RVP_I2S TGLH_0A70_HDA  TGLU_RVP_SDW  TGLU_VOLT_SDW TGLU_SKU0A3E_SDW TGLU_SKU0A32  ICL_RVP_HDA  JSL_RVP_NOCODEC CML_RVP_HDA CML_RVP_SDW CML_SKU0983_SDW

Can you please redo a test run with exactly the same settings as the daily test, only with your added firmware PR instead of the main branch? Thank you.

Edit: I meant the sof main branch + https://github.com/thesofproject/sof/pull/4386 and this PR https://github.com/thesofproject/linux/pull/2989 for the kernel

keyonjie commented:
> > @plbossart @kv2019i I added more platforms (includes the strict TGLU_RVP_NOCODEC) coverage and triggered a validation with combine this PR and the SOF [thesofproject/sof#4386](https://github.com/thesofproject/sof/pull/4386), the result looks quite good. Please refer to the CI test plan result 4814.
> 
> @keyonjie it would be good if your tests used the _same_ devices as the regular daily tests. You seem to have excellent results indeed, but added and removed platforms, so we can't do an apples-to-apples comparison.
> Missing platforms include:
> 
> TGLH_0A5E_SDW | EHL_RVP_I2S TGLH_0A70_HDA TGLU_RVP_SDW TGLU_VOLT_SDW TGLU_SKU0A3E_SDW TGLU_SKU0A32 ICL_RVP_HDA JSL_RVP_NOCODEC CML_RVP_HDA CML_RVP_SDW CML_SKU0983_SDW
> 
> Can you please redo a test run with exactly the same settings as the daily test, only with your added firmware PR instead of the main branch? Thank you.
> 
> Edit: I meant the sof main branch + [thesofproject/sof#4386](https://github.com/thesofproject/sof/pull/4386) and this PR #2989 for the kernel

yes, that just finished 2 hours back, please see the plan result 4819.

plbossart commented:
@keyonjie the results are difficult to interpret when comparing 
https://sof-ci.sh.intel.com/#/result/planresultdetail/4819
and 
https://sof-ci.sh.intel.com/#/result/planresultdetail/4803 (today's official daily test)

you've got one major issue with https://sof-ci.sh.intel.com/#/result/planresultdetail/4819?model=TGLU_RVP_NOCODEC&testcase=multiple-pause-resume-25

````
[ 4292.922751] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx: 0x60060000: GLB_STREAM_MSG: TRIG_PAUSE
[ 4292.923053] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: pcm: trigger stream 5 dir 1 cmd 3
[ 4293.430302] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: error: ipc timed out for 0x60060000 size 12
[ 4293.430320] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: info: preventing DSP entering D3 state to preserve context
[ 4293.430341] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: status: fw entered - code 00000005
[ 4293.430455] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: error: unexpected fault 0x00000000 trace 0x00004000
[ 4293.430474] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: error: hda irq intsts 0x00000000 intlctl 0xc0000083 rirb 00
[ 4293.430481] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: error: dsp irq ppsts 0x00000000 adspis 0x00000000
[ 4293.430496] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: error: host status 0x00000000 dsp status 0x00000000 mask 0x00000003
[ 4293.430503] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: error: waking up any trace sleepers
[ 4293.430532] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: FW Poll Status: reg[0x160]=0x140000 successful
[ 4293.430546] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: ASoC: error at soc_component_trigger on 0000:00:1f.3: -110
[ 4293.430556] kernel:  Port1: ASoC: trigger FE cmd: 3 failed: -110
[ 4293.430568] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: pcm: trigger stream 1 dir 0 cmd 4
[ 4293.430573] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: error: trace IO error
[ 4293.430607] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx: 0x60060000: GLB_STREAM_MSG: TRIG_PAUSE
[ 4293.431094] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: FW Poll Status: reg[0x160]=0x14001e successful
[ 4293.934276] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: error: ipc timed out for 0x60060000 size 12
[ 4293.934294] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: info: preventing DSP entering D3 state to preserve context
[ 4293.934304] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: status: fw entered - code 00000005
[ 4293.934428] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: error: unexpected fault 0x00000000 trace 0x00004000
[ 4293.934445] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: error: hda irq intsts 0x00000000 intlctl 0xc0000083 rirb 00
[ 4293.934452] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: error: dsp irq ppsts 0x00000000 adspis 0x00000000
[ 4293.934466] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: error: host status 0x00000000 dsp status 0x00000000 mask 0x00000003
[ 4293.934473] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: error: waking up any trace sleepers
[ 4293.934495] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: FW Poll Status: reg[0xa0]=0x20240000 successful
[ 4293.934506] kernel: sof-audio-pci-intel-tgl 0000:00:1f.3: ASoC: error at soc_component_trigger on 0000:00:1f.3: -110
[ 4293.934516] kernel:  DMIC16kHz: ASoC: trigger FE cmd: 3 failed: -110
````

You've got multiple platforms failing with multiple-pause-resume-25.sh, and two issues with TGLU_VOLT_SDW that are supposedly fixed in the daily test.

And last you've got a kernel oops here:
https://sof-ci.sh.intel.com/#/result/planresultdetail/4819?model=CML_HEL_RT5682&testcase=check-kmod-load-unload-25

I really don't know how to draw a conclusion, it's like the tests were done on the PRs themselves instead of the PR applied to the main branch?

keyonjie commented:
@plbossart yes, the CI just uses the last built binaries directly for the PR plan test, no applying on top of latest main happen, that's why we don't see TGLU_VOLT_SDW xrun fixes with it.

The TGLU_RVP_NOCODEC xrun issue happens in daily randomly these days, I will try to analyze that today first.

The Helios module unload/reload issue will be look into next.

RanderWang reviewed (dismissed):
LGTM. we need to add this flag for IPC4

keyonjie commented:
SOFCI TEST

keyonjie commented:
SOFCI TEST

keyonjie commented:
This need to go after FW PR https://github.com/thesofproject/sof/pull/4386
please check the combined CI result in the test planresult 7341.

plbossart commented on sound/soc/sof/intel/hda-loader.c:
```diff
@@ -371,6 +371,40 @@ int hda_dsp_cl_boot_firmware_iccmax(struct snd_sof_dev *sdev)
+	sdev->enabled_cores_mask |= chip->init_core_mask;
+	mask = sdev->enabled_cores_mask;
+	for_each_set_bit(j, &mask, SOF_MAX_DSP_NUM_CORES)
+		sdev->dsp_core_ref_count[j]++;
```
@keyonjie I was not able to understand the difference with the first boot where this code is used

````c
/* step 1: power up corex */
ret = hda_dsp_enable_core(sdev, chip->host_managed_cores_mask);
````

init_core_mask, enable_cores_mask, host_managed_cores_mask, that looks like a recipe for bugs.

keyonjie replied:
> @keyonjie I was not able to understand the difference with the first boot where this code is used
> 
> ```c
> /* step 1: power up corex */
> ret = hda_dsp_enable_core(sdev, chip->host_managed_cores_mask);
> ```
> 
> init_core_mask, enable_cores_mask, host_managed_cores_mask, that looks like a recipe for bugs.

It could be, but that is not introduced by this PR, no?

ranj063 commented on sound/soc/sof/intel/hda-loader.c:
```diff
@@ -371,6 +371,40 @@ int hda_dsp_cl_boot_firmware_iccmax(struct snd_sof_dev *sdev)
+	int ret;
+
+	/* init for booting wait */
+	init_waitqueue_head(&sdev->boot_wait);
```
this is not needed. In fact it should be removed from hda_dsp_cl_boot_firmware() as well. snd_sof_run_firmware() already does it

keyonjie replied:
> this is not needed. In fact it should be removed from hda_dsp_cl_boot_firmware() as well. snd_sof_run_firmware() already does it

What you mentioned could be true, but that should belong to another PR? Here the PR is just following what the hda_dsp_cl_boot_firmware() does.

ranj063 replied:
no I think we should remove it from here. We can remoe it from hda_dsp_cl_boot_firmware() in another PR

ranj063 commented on sound/soc/sof/intel/hda-loader.c:
```diff
@@ -371,6 +371,40 @@ int hda_dsp_cl_boot_firmware_iccmax(struct snd_sof_dev *sdev)
+	}
+
+	/* set enabled cores mask and increment ref count for cores in init_core_mask */
+	sdev->enabled_cores_mask |= chip->init_core_mask;
```
@keyonjie paranoia check here: I hope you have checked that this is OK to do i.e just power up the init_core_mask. When we download the fw, we power up all cores and then power down ~init_core_mask cores.

keyonjie replied:
> @keyonjie paranoia check here: I hope you have checked that this is OK to do i.e just power up the init_core_mask. When we download the fw, we power up all cores and then power down ~init_core_mask cores.

Yes, the CI test showed me that we actually don't need to power up other Cores not covered by init_core_mask.

plbossart replied:
@keyonjie you should base your results on the hardware recommended sequences, not experimental results from a small set of devices in the CI pool. That really doesn't inspire confidence if I'm honest.

keyonjie replied:
> @keyonjie you should base your results on the hardware recommended sequences, not experimental results from a small set of devices in the CI pool. That really doesn't inspire confidence if I'm honest.

That's a good point, I didn't find any hardware-recommended sequences mentioning that we have to power on all DSP cores at the booting, @ranj063 did you?

ranj063 commented:
@keyonjie could you please run a daily test with this PR?

keyonjie commented:
@ranj063 @plbossart looks it doesn't build anymore with simple rebasing, will work on it later.

ranj063 commented:
> @ranj063 @plbossart looks it doesn't build anymore with simple rebasing, will work on it later.

@keyonjie let us know when this is ready for review!

keyonjie commented:
@ranj063 @plbossart ready for review now.

keyonjie commented:
@ranj063 @plbossart @lgirdwood to test IMR restore feature, we need to config CONFIG_CAVS_IMR_D3_PERSISTENT for each platforms intended to use that, the CI test report is here:
https://sof-ci.sh.intel.com/#/result/planresultdetail/8827

you will see dmesg log "IMR restore supported, booting from IMR directly" which means IMR restore is used there.

lgirdwood commented:
@mengdonglin are we able to test with this config from Keyon after v2/v1.9.3 validation is complete ?

ranj063 commented on include/sound/sof/info.h:
```diff
@@ -25,6 +25,7 @@
 #define SOF_IPC_INFO_LOCKS		BIT(1)
 #define SOF_IPC_INFO_LOCKSV		BIT(2)
 #define SOF_IPC_INFO_GDB		BIT(3)
+#define SOF_IPC_INFO_D3_PERSISTENT	BIT(4)
```
Can you also fix the typos in the commit message: "Add a bit to denote if the FW supports the IMR..." and Firmware in all small-case in the second line.

ranj063 commented on sound/soc/sof/intel/hda-loader.c:
```diff
@@ -25,6 +25,44 @@
 
 #define HDA_CL_STREAM_FORMAT 0x40
 
+static void hda_set_ssps_slave(struct snd_sof_dev *sdev)
```
typos in commit message: "Move the SSP DAI's slave mode set function to the hda_set_ssps_slave() helper and the SDW wakeen interrupt enabling to the hda_sdw_wakeen_resume() helper"

ranj063 commented on sound/soc/sof/intel/hda-loader.c:
```diff
@@ -25,6 +25,44 @@
 
 #define HDA_CL_STREAM_FORMAT 0x40
 
+static void hda_set_ssps_slave(struct snd_sof_dev *sdev)
```
maybe call this function hda_ssp_resume()? then it will avoid using "slave" and it will leave the door open to move other things into it

keyonjie replied:
Good suggestion, let me change them.

ranj063 commented on sound/soc/sof/intel/hda-loader.c:
```diff
@@ -25,6 +25,44 @@
+	}
+}
+
+static void hda_sdw_wakeen_resume(struct snd_sof_dev *sdev)
```
same here drop the wakeen from the function name

ranj063 commented on sound/soc/sof/intel/hda-loader.c:
```diff
@@ -374,6 +374,40 @@ int hda_dsp_cl_boot_firmware_iccmax(struct snd_sof_dev *sdev)
+	/* power up & unstall/run the cores to run the firmware */
+	ret = hda_dsp_enable_core(sdev, chip->init_core_mask);
+	if (ret < 0) {
+		dev_err(sdev->dev, "error: dsp core start failed %d\n", ret);
```
remover error prefix

keyonjie replied:
yes, it could be superfluous, but aligns to all error prints in this file?

ranj063 replied:
no we're slowly moving away from the error prefix. so no new ones please

keyonjie replied:
fair, updated now.

ranj063 commented on sound/soc/sof/intel/hda-loader.c:
```diff
@@ -374,6 +374,40 @@ int hda_dsp_cl_boot_firmware_iccmax(struct snd_sof_dev *sdev)
 	return ret;
 }
 
+static int hda_dsp_boot_imr(struct snd_sof_dev *sdev)
```
you call it IMR restore support with no mention on it in the code anywhere. The code has only references to D3_PERSISTENT. Can we just stick to that in the commit messages as well so that reader don't get confused. Also I think PERSISTENT_D3 sounds more intuitive than D3_PERSISTENT. D3_PERSISTENT sounds like the D3 state will be persistent.

ranj063 replied:
acutally if I read back my own comment, both of them convey the same message. Maybe we should call it IMR_PERSISTENT_D3 or something similar to say whats persistent in D3

keyonjie replied:
Per my understanding, the IMR D3 persistency is our system behavior and should not be configurable, but either to use the IMR restoring feature or not is configurable.

Yes, the naming SOF_IPC_INFO_D3_PERSISTENT is somewhat confusing with respect to what I mentioned above, it should be SOF_IPC_INFO_IMR_RESTORE_SUPPORT, but since the naming is already quite long, and it needs to align to the existed FW side also, I would suggest keeping it at the moment and change it later from both side if needed.

keyonjie commented:
the latest feature test result here: https://sof-ci.sh.intel.com/#/result/planresultdetail/8877

ranj063 commented on sound/soc/sof/intel/hda-loader.c:
```diff
@@ -416,6 +417,7 @@ int hda_dsp_cl_boot_firmware(struct snd_sof_dev *sdev)
 	int ret, ret1, i;
 
 	if ((sdev->fw_ready.flags & SOF_IPC_INFO_D3_PERSISTENT) &&
+	    !(sof_debug_check_flag(SOF_DBG_IGNORE_D3_PERSISTENT)) &&
```
this patch has to be squashed with the previous one no? Because we'd enable it by default in the previous one and only disable it based on the module param in this one

keyonjie replied:
my design here is to enable it by default if the FW claims it supports the feature. So this subsequent module param is only adding an interface to disable using the FW feature on the host side, for debugging purposes only.

ranj063 reviewed (approved):
looks good @keyonjie. I'd like the lat 2 patches to still be squashed but I'll leave that for others to comment.

plbossart commented on sound/soc/sof/intel/hda-loader.c:
```diff
@@ -398,23 +427,7 @@ int hda_dsp_cl_boot_firmware(struct snd_sof_dev *sdev)
-	 */
-	if (!sdev->first_boot)
-		hda_sdw_process_wakeen(sdev);
+	hda_sdw_resume(sdev);
```
we could move the !sdev->first_boot inside the helper, no need to add a new definition that's rather misleading.

keyonjie replied:
> we could move the !sdev->first_boot inside the helper, no need to add a new definition that's rather misleading.

checking sdev->first_boot inside hda_sdw_process_wakeen() sounds a bit odd to me, @plbossart do you really want that?

plbossart replied:
I actually don't see a need for a new wrapper here, just use 	

if (!sdev->first_boot)
		hda_sdw_process_wakeen(sdev);

twice...

plbossart replied:
actually the imr stuff only happens after the first boot so the test is not needed ... really no need for a wrapper.

plbossart commented on sound/soc/sof/intel/hda-loader.c:
```diff
@@ -101,15 +138,7 @@ static int cl_dsp_init(struct snd_sof_dev *sdev, int stream_tag)
-						 SSP_SET_SLAVE,
-						 SSP_SET_SLAVE);
-	}
+	hda_ssp_resume(sdev);
```
the commit message mentions a different name so that needs to be modified.

what was wrong with  hda_set_ssps_slave()?

ranj063 replied:
not inclusive

plbossart replied:
this calls SSP_SET_SLAVE...

I am all for inclusive terms, but this has nothing to do with resume operations.

keyonjie replied:
> this calls SSP_SET_SLAVE...
> 
> I am all for inclusive terms, but this has nothing to do with resume operations.

I think Ranjani means "what we need to perform to SSPs when resuming audio device (D3->D0)", like "_ssp_at_resume()". And, use the more generic _ssp_resume() will be friendly to add more SSP related reset/resume/init stuff into it in future when needed.

I am fine on both, just let me know the aligned one.

ranj063 replied:
just because we do set_slave doesnt mean we have to name the function that too no? Someday maybe set_slave will get renamed and we wont have to worry about the static function name having the word"slave" in it. But I agree resume is a bad name. This is the loader, so while it is also called during resume, thats not the only thing. Why not call is something lika hda_ssp_cl_init() and hda_sdw_cl_init()

plbossart replied:
the acronym we came up with is CBP Codec BitClockProvider, so we can call this wrapper

hda_ssp_set_cbp()

if it also sets the fsync hda_ssp_set_cbp_cfp().

and no wrapper for SoundWire (see comment below) would be my preference.

plbossart replied:
sound/soc/sof/intel/hda-loader.c:                                                SSP_SET_SLAVE,
sound/soc/sof/intel/hda-loader.c:                                                SSP_SET_SLAVE);
sound/soc/sof/intel/hda.h:#define SSP_SET_SLAVE         (SSP_SET_SCLK_SLAVE | SSP_SET_SFRM_SLAVE)

This should be SSP_SET_CBP_CFP

plbossart commented on sound/soc/sof/sof-priv.h:
```diff
@@ -35,6 +35,9 @@
 							*/
 #define SOF_DBG_PRINT_ALL_DUMPS		BIT(6) /* Print all ipc and dsp dumps */
+#define SOF_DBG_IGNORE_D3_PERSISTENT		BIT(7) /* ignore the DSP D3 persistent capability
+							* and always do FW downloading at resume
```
and always download firmware upon D3 exit.

plbossart reviewed (changes requested):
minor comments below, the names for the new helpers are not very good and not aligned with the commit messages.

plbossart commented:
I corrected my own comments in PR #3340, let's go from there. Thanks @keyonjie and fair winds to you.
